svcb.go

```diff
@@ -0,0 +1,795 @@
+}
+
+func (s *SVCBNoDefaultAlpn) parse(b string) error {
+	if len(b) != 0 {
+		return errors.New("dns: no_default_alpn should have no value")
+	}
+	return nil
+}
+
+// SVCBPort pair defines the port for connection.
+// Basic use pattern for creating a port option:
+//
+//	o := new(dns.SVCB)
+//	o.Hdr.Name = "."
+//	o.Hdr.Rrtype = dns.SVCB
+//	e := new(dns.SVCBPort)
+//	e.Port = 80
+//	o.Value = append(o.Value, e)
+type SVCBPort struct {
+	Port uint16
+}
+
+func (s *SVCBPort) Key() uint16        { return SVCB_PORT }
+func (s *SVCBPort) String() string     { return strconv.FormatUint(uint64(s.Port), 10) }
+func (s *SVCBPort) copy() SVCBKeyValue { return &SVCBPort{s.Port} }
+func (s *SVCBPort) len() uint16        { return 2 }
+
+func (s *SVCBPort) unpack(b []byte) error {
+	if len(b) != 2 {
+		return errors.New("dns: bad port")
+	}
+	s.Port = binary.BigEndian.Uint16(b)
+	return nil
+}
+
+func (s *SVCBPort) pack() ([]byte, error) {
+	b := make([]byte, 2)
+	binary.BigEndian.PutUint16(b, s.Port)
+	return b, nil
+}
+
+func (s *SVCBPort) parse(b string) error {
+	port, err := strconv.ParseUint(b, 10, 16)
+	if err != nil {
+		return errors.New("dns: bad port")
+	}
+	s.Port = uint16(port)
+	return nil
+}
+
+// SVCBIPv4Hint pair suggests an IPv4 address
+// which may be used to open connections if A and AAAA record
+// responses for SVCB's Target domain haven't been received.
+// In that case, optionally, A and AAAA requests can be made,
+// after which the connection to the hinted IP address may be
+// terminated and a new connection may be opened.
+// Basic use pattern for creating an ipv4hint option:
+//
+//	o := new(dns.HTTPS)
+//	o.Hdr.Name = "."
+//	o.Hdr.Rrtype = dns.HTTPS
+//	e := new(dns.SVCBIPv4Hint)
+//	e.Hint = []net.IP{net.IPv4(1,1,1,1).To4()}
+//  // or
+//	e.Hint = []net.IP{net.ParseIP("1.1.1.1").To4()}
+//	o.Value = append(o.Value, e)
+type SVCBIPv4Hint struct {
+	Hint []net.IP // Always IPv4
+}
+
+func (s *SVCBIPv4Hint) Key() uint16 { return SVCB_IPV4HINT }
+func (s *SVCBIPv4Hint) len() uint16 { return uint16(4 * len(s.Hint)) }
+
+func (s *SVCBIPv4Hint) pack() ([]byte, error) {
+	b := make([]byte, 0, 4*len(s.Hint))
+	for _, e := range s.Hint {
+		x := e.To4()
+		if x == nil {
+			return nil, errors.New("dns: not IPv4")
+		}
+		b = append(b, x...)
+	}
+	return b, nil
+}
+
+func (s *SVCBIPv4Hint) unpack(b []byte) error {
+	if len(b) == 0 || len(b)%4 != 0 {
+		return errors.New("dns: bad array of IPv4 addresses")
+	}
+	i := 0
+	x := make([]net.IP, 0, len(b)/4)
+	for i < len(b) {
+		x = append(x, net.IP(b[i:i+4]))
+		i += 4
+	}
+	s.Hint = x
+	return nil
+}
+
+// String returns "<nil>" if an invalid IPv4 address was encountered.
+// TODO DOC Do I need full definition for doc?
+func (s *SVCBIPv4Hint) String() string {
+	var str strings.Builder
+	str.Grow(16 * len(s.Hint))
+	for _, e := range s.Hint {
+		x := e.To4()
+		if x == nil {
+			return "<nil>"
+		}
+		str.WriteByte(',')
+		str.WriteString(x.String())
+	}
+	return str.String()[1:]
```

Comment:
Why not just use `strings.Join` and an approach similar to `(*SVCBMandatory).String` here?

Reply:
In this case we know the maximum length of each string which is also a good approximation for length while the strings approach uses a lot of pointers and much less efficient. Anyways, no one will benchmark these mini-functions so readability is more important.